kusto/kcsb.go

```diff
@@ -0,0 +1,428 @@
+	"application certificate": applicationCertificate, "applicationcertificate": applicationCertificate,
+	"application certificate thumbprint": applicationCertificateThumbprint, "applicationcertificatethumbprint": applicationCertificateThumbprint,
+	"sendcertificatechain": sendCertificateChain, "send certificate chain": sendCertificateChain,
+	"authority id": authorityId, "authorityid": authorityId, "authority": authorityId, "tenantid": authorityId, "tenant": authorityId, "tid": authorityId,
+	"application token": applicationToken, "applicationtoken": applicationToken, "apptoken": applicationToken,
+	"user token": userToken, "usertoken": userToken, "usrtoken": userToken,
+	"msi_auth":               msiAuth,
+	"managedserviceidentity": managedServiceIdentity, "managed service identity": managedServiceIdentity,
+	"interactive login": interactiveLogin, "interactivelogin": interactiveLogin,
+	"az cli": azCli, "azcli": azCli,
+	"domain hint": domainHint, "domainhint": domainHint,
+}
+
+func requireNonEmpty(key string, value string) {
+	if isEmpty(value) {
+		panic(fmt.Sprintf("Error: %s cannot be null", key))
+	}
+}
+
+func assignValue(kcsb *ConnectionStringBuilder, rawKey string, value string) error {
+	rawKey = strings.ToLower(strings.Trim(rawKey, " "))
+	parsedKey, ok := csMapping[rawKey]
+	if !ok {
+		return fmt.Errorf("Error: unsupported key %q in connection string ", rawKey)
+	}
+	switch parsedKey {
+	case dataSource:
+		kcsb.DataSource = value
+	case aadUserId:
+		kcsb.AadUserID = value
+	case password:
+		kcsb.Password = value
+	case applicationClientId:
+		kcsb.ApplicationClientId = value
+	case applicationKey:
+		kcsb.ApplicationKey = value
+	case applicationCertificate:
+		kcsb.ApplicationCertificate = value
+	case applicationCertificateThumbprint:
+		kcsb.ApplicationCertificateThumbprint = value
+	case sendCertificateChain:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.SendCertificateChain = bval
+	case authorityId:
+		kcsb.AuthorityId = value
+	case applicationToken:
+		kcsb.ApplicationToken = value
+	case userToken:
+		kcsb.UserToken = value
+	case msiAuth:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.MsiAuthentication = bval
+	case managedServiceIdentity:
+		kcsb.ManagedServiceIdentity = value
+	case azCli:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.Azcli = bval
+	case interactiveLogin:
+		bval, _ := strconv.ParseBool(value)
+		kcsb.InteractiveLogin = bval
+	case domainHint:
+		kcsb.RedirectURL = value
+	}
+	return nil
+}
+
+/*
+Creates new KustoConnectionStringBuilder.
+Params takes kusto connection string connStr: string.  Kusto connection string should be of the format:
+https://<clusterName>.kusto.windows.net;AAD User ID="user@microsoft.com";Password=P@ssWord
+For more information please look at:
+https://docs.microsoft.com/azure/data-explorer/kusto/api/connection-strings/kusto
+*/
+func GetConnectionStringBuilder(connStr string) *ConnectionStringBuilder {
```

Comment:
@yogilad Kcsb is in Kusto package, and the `New(...)` function is already being used. Should we create a new package just for `ConnectionStringBuilder` ? So that we can achieve ConnectionStringBuilder.New(...)

Intentions here is to use like:
* _If created with just string keywords_
 kusto.GetConnectionStringBuilder(...) 


* _If created using dedicated With*** apis_
{kcsb obj}.With***(...)

Reply:
I see I made this comment earlier (sorry).
If it stays in the same module, it should be **New**ConnectionStringBuilder.